src/landpatterns/pad-planner.stanza

```diff
@@ -168,30 +198,50 @@ public defmethod th-pad-generator (x:PadPlanner, row:Int, column:Int) -> ( (Dims
         npth-pad(h(hole))
       npth-gen
     (a:True, f, h):
-      ; TODO - Log Warning Here ?
-      ; this is a exceptional case but I'm not sure it is an error
-      false
+      throw $ NoGeneratorForActivePadError(row, column, "shape- or hole-")
 
 doc: \<DOC>
 Hole Shape for Pad
 This function provides the user with the ability to create a
 plated through-hole pad instead of an SMT pad.
 
+In general, this function should not query `active?` unless
+it needs to. The `active?` flag will be queried in the `pad-generator`
+functions when selecting whether to generate pad shapes.
+
+The typical use case is to only return `False` if the land pattern
+contains zero through-hole pads. If you need to create a land pattern
+with a mix of through-hole and surface mount pads - it is better to construct
+this land pattern through composition.
+
+If you intend to violate this convention - you must provide a reformed
+version of {@link is-through-hole?} in your derived implementation.
+
 @param x Pad Planner (self)
 @param row Zero-based Index into the grid of a package's pad positions.
 @param column Zero-based Index into the grid of a package's pad positions.
 
-@return If this function returns `false` then no hole is created.
-Otherwise, if a the function returns a shape, then holes of that
-shape will be created. Note - that you can pass any shape here but
-your fabricator may have limitations on what shapes it can create.
+@return This function should return a function that will convert
+a `Dims` object into a `Shape`. That function will need to have the
+following signature:
+
+```stanza
+Dims -> Shape
+```
+
+This function can return `false` for location in the grid where no
+hole should be created. If this function returns `false` at an
+`active? = true` grid location - then the related pad generation
+functions will raise an exception.
 
-TODO - Should we make this return `Circle|Capsule` instead of `Shape` ?
-  Holes in the board generally speaking are very expensive to make in
-  other shapes but I don't necessarily want to limit things here.
-  Perhaps with make non-Circle/Capsule shaped holes a check that is
-  disable-able so that the user can override.
 <DOC>
+
+; TODO - Should we make this return `Circle|Capsule` instead of `Shape` ?
+;   Holes in the board generally speaking are very expensive to make in
```

Comment:
If the enterprise customers can afford it then I would  suggest keep it `Shape`

Reply:
We can keep the interface as `Shape`.